manifest/platform_matcher.go

```diff
@@ -0,0 +1,120 @@
+package manifest
+
+// Largely based on
+// https://github.com/moby/moby/blob/bc846d2e8fe5538220e0c31e9d0e8446f6fbc022/distribution/cpuinfo_unix.go
+// https://github.com/containerd/containerd/blob/726dcaea50883e51b2ec6db13caff0e7936b711d/platforms/cpuinfo.go
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/containers/image/v5/types"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// For Linux, the kernel has already detected the ABI, ISA and Features.
+// So we don't need to access the ARM registers to detect platform information
+// by ourselves. We can just parse these information from /proc/cpuinfo
+func getCPUInfo(pattern string) (info string, err error) {
+	if runtime.GOOS != "linux" {
+		return "", fmt.Errorf("getCPUInfo for OS %s not implemented", runtime.GOOS)
+	}
+
+	cpuinfo, err := os.Open("/proc/cpuinfo")
+	if err != nil {
+		return "", err
+	}
+	defer cpuinfo.Close()
+
+	// Start to Parse the Cpuinfo line by line. For SMP SoC, we parse
+	// the first core is enough.
+	scanner := bufio.NewScanner(cpuinfo)
+	for scanner.Scan() {
+		newline := scanner.Text()
+		list := strings.Split(newline, ":")
+
+		if len(list) > 1 && strings.EqualFold(strings.TrimSpace(list[0]), pattern) {
+			return strings.TrimSpace(list[1]), nil
+		}
+	}
+
+	// Check whether the scanner encountered errors
+	err = scanner.Err()
+	if err != nil {
+		return "", err
+	}
+
+	return "", fmt.Errorf("getCPUInfo for pattern: %s not found", pattern)
+}
+
+func getCPUVariant() string {
+	if runtime.GOOS == "windows" {
+		// Windows only supports v7 for ARM32 and v8 for ARM64 and so we can use
+		// runtime.GOARCH to determine the variants
+		var variant string
+		switch runtime.GOARCH {
+		case "arm64":
+			variant = "v8"
+		case "arm":
+			variant = "v7"
+		default:
+			variant = "unknown"
+		}
+
+		return variant
+	}
+
+	variant, err := getCPUInfo("Cpu architecture")
+	if err != nil {
+		return ""
+	}
+	// TODO handle RPi Zero mismatch (https://github.com/moby/moby/pull/36121#issuecomment-398328286)
+
+	switch strings.ToLower(variant) {
+	case "8", "aarch64":
+		variant = "v8"
+	case "7", "7m", "?(12)", "?(13)", "?(14)", "?(15)", "?(16)", "?(17)":
+		variant = "v7"
+	case "6", "6tej":
+		variant = "v6"
+	case "5", "5t", "5te", "5tej":
+		variant = "v5"
+	case "4", "4t":
+		variant = "v4"
+	case "3":
+		variant = "v3"
+	default:
+		variant = "unknown"
+	}
+
+	return variant
+}
+
+func WantedPlatform(ctx *types.SystemContext) (imgspecv1.Platform, error) {
+	wantedArch := runtime.GOARCH
+	if ctx != nil && ctx.ArchitectureChoice != "" {
+		wantedArch = ctx.ArchitectureChoice
+	}
+	wantedOS := runtime.GOOS
+	if ctx != nil && ctx.OSChoice != "" {
+		wantedOS = ctx.OSChoice
+	}
+
+	wantedPlatform := imgspecv1.Platform{
+		OS:           wantedOS,
+		Architecture: wantedArch,
+	}
+
+	if wantedArch == "arm" || wantedArch == "arm64" {
+		if ctx != nil && ctx.VariantChoice != "" {
```

Comment:
Why not switch the nesting of these two `if`, so that user input is always used, and guessing happens only on ARM? Silently ignoring user input would be surprising.

Reply:
Sounds reasonable.

Reply:
Oh, actually no. I think that we need to always guess variant on arm if no user input is specified to download the proper variant. On any other architecture, variant does not seem to be used, so it should be ignored or nulled. Or maybe fail if there is user input for `VariantChoice`.

Reply:
That `variant` is only used on ARM is just the current state, not an inherent property of the universe.

Allowing `VariantChoice` on other architectures could, in the future, allow using this version of the code (unaware of variants on those architectures) by newer callers (aware of variants), explicitly making a choice that c/image cannot (yet) make.

> Or maybe fail if there is user input for VariantChoice.

Yes; it should _not_ be ignored (because, again, in some future version it may start to matter, and code writing nonsense in there would start to fail).

Given an architecture that does not have variants, and an user-provided `VariantChoice`, I would expect the architecture-independent part of this PR to _naturally_, without special-casing, reject any image — exactly the same as if the caller asked for ARM variant `this-is-not-an-ARM-variant-at-all`.

Reply:
> reject any image

reject any image that does not have such a variant explicitly present, that is.

Reply:
This approach totally makes sense. I'll see what I can do.